Examples/Physics_applications/capacitive_discharge/PICMI_inputs_2d.py

```diff
@@ -45,6 +53,161 @@
+
+        super(PoissonSolverPseudo1D, self).initialize_inputs()
+
+        self.nx = self.grid.nx
+        self.nz = self.grid.ny
+        self.dx = (self.grid.xmax - self.grid.xmin) / self.nx
+        self.dz = (self.grid.ymax - self.grid.ymin) / self.nz
+
+        if not np.isclose(self.dx, self.dz):
+            raise RuntimeError('Direct solver requires dx = dz.')
+
+        self.nxguardrho = 2
+        self.nzguardrho = 2
+        self.nxguardphi = 1
+        self.nzguardphi = 1
+
+        self.phi = np.zeros(
+            (self.nx + 1 + 2*self.nxguardphi,
+            self.nz + 1 + 2*self.nzguardphi)
+        )
+
+        self.decompose_matrix()
+
+        callbacks.installpoissonsolver(self._run_solve)
+
+    def decompose_matrix(self):
+        """Function to build the superLU object used to solve the linear
+        system."""
+        self.nxsolve = self.nx + 1
+        self.nzsolve = self.nz + 3
+
+        # Set up the computation matrix in order to solve A*phi = rho
+        A = np.zeros(
+            (self.nzsolve*self.nxsolve, self.nzsolve*self.nxsolve)
+        )
+        kk = 0
+        for ii in range(self.nxsolve):
+            for jj in range(self.nzsolve):
+                temp = np.zeros((self.nxsolve, self.nzsolve))
+
+                if ii == 0 or ii == self.nxsolve - 1:
+                    temp[ii, jj] = 1.
+                elif ii == 1:
+                    temp[ii, jj] = -2.0
+                    temp[ii-1, jj] = 1.0
+                    temp[ii+1, jj] = 1.0
+                elif ii == self.nxsolve - 2:
+                    temp[ii, jj] = -2.0
+                    temp[ii+1, jj] = 1.0
+                    temp[ii-1, jj] = 1.0
+                elif jj == 0:
+                    temp[ii, jj] = 1.0
+                    temp[ii, -3] = -1.0
+                elif jj == self.nzsolve - 1:
+                    temp[ii, jj] = 1.0
+                    temp[ii, 2] = -1.0
+                else:
+                    temp[ii, jj] = -4.0
+                    temp[ii, jj+1] = 1.0
+                    temp[ii, jj-1] = 1.0
+                    temp[ii-1, jj] = 1.0
+                    temp[ii+1, jj] = 1.0
+
+                A[kk] = temp.flatten()
+                kk += 1
+
+        A = csc_matrix(A, dtype=np.float32)
+        self.lu = sla.splu(A)
+
+    def _run_solve(self):
+        """Function run on every step to perform the required steps to solve
+        Poisson's equation."""
+
+        # get rho from WarpX
+        if self.rho_wrapper is None:
+            self.rho_wrapper = fields.RhoFPWrapper(0, True)
+        self.rho_data = self.rho_wrapper[Ellipsis][:,:,0]
+
+        self.solve()
+
+        if self.phi_wrapper is None:
+            self.phi_wrapper = fields.PhiFPWrapper(0, True)
+        self.phi_wrapper[Ellipsis] = self.phi
+
+    def solve(self):
+        """The solution step. Includes getting the boundary potentials and
+        calculating phi from rho."""
+        t = _libwarpx.libwarpx.warpx_gett_new(0)
```

Comment:
It is used in the `right_voltage = eval(self.right_voltage)` call below since `self.right_voltage` is a string function of `t`. I did it this way so that the input script can easily be switched back-and-forth between using the MLMG solver and the direct solver. Do you have a suggestion of a way to evaluate an expression in a PICMI file that would solve this issue?

Reply:
You can explicitly pass in a dictionary, `right_voltage = eval(self.right_voltage, {'t':_libwarpx.libwarpx.warpx_gett_new(0)})`. This would be more concise since it makes the dependence on `t` explicit and avoids using `globals`.

Reply:
Thanks! That is a much better alternative.